### ISSUE TITLE
Limit number of concurrent substeps.

### ISSUE THREAD

**Issue by BoPeng** [vatlab/sos#1296]
Per communication with a gitter user, he has

```
[10]
substeps with high demand

[20]
substeps with lower demand
```
when the workflow is executed with `-j  20`, step 10 would quit with out of memeory, but if the workflow is executed with `-j 10`, the machine would be under utilized for step 20.

My reply:
> Unfortunately not at this point. Right now all workers accept steps, substeps, and subworkflows whenever they are available, regardless of their demand on resources so the high-demand steps would be sent to all available workers.
> I see there are two choices: one is to specify the resource needed for each substep so that workers would not execute a substep when there is no enough resource. But this is hard because for example, the machine might have enough resource when the substep is submitted.
> Another possibility is to expand option concurrent, which accept `True/False` for sending substeps concurrently. It could accept a number, e.g. `concurrent=8` so that step 10 would send at most 8 substeps to workers at a time. A mechanism like this is actually already in place to keep the number of running substeps below 100 to avoid overwhelming the controller with too many substeps.

**Comment by gaow** [comment]
The 2nd option sounds good! But can we also achieve it with `task` using the options it currently has?

**Comment by BoPeng** [comment]
It looks like that `concurrent=8` is the easiest solution but it hard-codes a constraint for a particular machine, which might not be desired. Perhaps something like **my substep will take 3 slots of concurrent workload**. So we have a few choices:

Assuming `-j 20`:

1. `concurrent=8`: the step will release at most `8` substeps to the workers, recognizing that the machine can only handle 8 concurrent substeps. However, in a complex scenario the 12 other workers will also be busy, working on something else and trigger out of memory.

2. `concurrent=3`: each step will occupy 3 workers, meaning that if at most 6 substeps will be executing, and 2 will still be working, and 12 workers will be idle.

**Comment by gaow** [comment]
> but it hard-codes a constraint for a particular machine, which might not be desired.

yes that's why I was thinking it would be easier if we can simply achieve this advanced usage via machine specific templates (`max_job` option). If that is possible, I dont think we should introduce additional complexities here.

**Comment by BoPeng** [comment]
You are right, if we do
```
[10]
task:
substeps with high demand

[20]
substeps with lower demand
```
then we can do
```
sos run -q localhost -J8 -j 20
```
to run 8 concurrent tasks for high-demand substeps, and 20 workers for low-demand substeps.

**Comment by BoPeng** [comment]
I have implemented the `concurrent=INT` behavior because it is a very natural extension of what we have right now.

With the patch and the following workflow
```
import time
st = time.time()

[1]
input: for_each=dict(i=range(10)), concurrent=2
print(f'Step1/{i}: start at {time.time() - st:.1f}')
time.sleep(5)

[2]
input: for_each=dict(i=range(10))
print(f'Step 2/{i}: start at {time.time() - st:.1f}')
time.sleep(5)
```
the output of command
```
 sos run test -v0 -j10
```
is
```
$ sos run test -v0 -j10
Step1/0: start at 0.1
Step1/1: start at 0.2
Step1/2: start at 5.2
Step1/3: start at 5.2
Step1/4: start at 10.3
Step1/5: start at 10.3
Step1/6: start at 15.3
Step1/7: start at 15.4
Step1/8: start at 20.4
Step1/9: start at 20.4
Step 2/0: start at 25.6
Step 2/1: start at 25.6
Step 2/3: start at 25.6
Step 2/2: start at 25.6
Step 2/4: start at 25.6
Step 2/5: start at 25.6
Step 2/7: start at 25.6
Step 2/6: start at 25.6
Step 2/8: start at 25.6
Step 2/9: start at 25.7
```
